src/Stratis.Bitcoin.Features.MemoryPool/MempoolValidator.cs

```diff
@@ -392,17 +399,33 @@ protected virtual void PreMempoolChecks(MempoolValidationContext context)
                 context.State.Fail(MempoolErrors.Coinstake).Throw();
             }
 
-            bool witnessEnabled = false;
+            // Reject transactions with witness data before segregated witness activates.
+            // https://github.com/bitcoin/bitcoin/blob/ea729d55b4dbd17a53ced474a8457d4759cfb5a5/src/validation.cpp#L463-L467
+
+            // TODO: The Bitcoin Core approach to segwit has changed since the MempoolValidator was first written; it is now regarded as always active
+            // and the following check has been removed. However, as we have not deployed segwit yet we should retain this check for now. It can
+            // be removed at a later stage & the witnessEnabled flags no longer used.
+            // https://github.com/bitcoin/bitcoin/commit/fa7a6cf1b36284db70e941bd2915fd6edbb0f9d6?diff=split#diff-24efdb00bfbe56b140fb006b562cc70bL581
+
+            // Validation of the actual witness data, if applicable, is done later when signature verification is performed.
+
+            DeploymentFlags flags = this.nodeDeployments.GetFlags(this.chainIndexer.Tip);
+            bool witnessEnabled = flags.ScriptFlags.HasFlag(ScriptVerify.Witness);
+
+            if (context.Transaction.HasWitness && !witnessEnabled)
```

Comment:
I wonder if this could be temporary cached at class level, if 100 trx arrive in a few seconds this will not change, so a simple comparison to current height will determine if witnessEnabled is still false (once its true no need to check it again)

Reply:
Something like this
```
if(!this.witnessEnabled && this.LastCheckHeight != this.chainIndexer.Tip.Height)
{
  this.LastCheckHeight = CI.Height
  this.witnessEnabled = // check flag
}
```

Reply:
It might make sense to do it down in the `NodeDeployments` instead so that all features have it cached - perhaps this could be a future improvement, depending on the expense of the `GetFlags()` call.